### PR TITLE
SIGINT-2058 - Added code for SRM scan (Freestyle, Pipeline, Multibranch)

### PR DESCRIPTION
**Pull request by sadmananik**

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle.java [hunk 1]
```diff
@@ -733,6 +835,76 @@ public void setMark_build_status(String mark_build_status) {
         this.mark_build_status = mark_build_status;
     }
 
+    @DataBoundSetter
+    public void setSrm_url(String srm_url) {
+        this.srm_url = srm_url;
```
Comment (zaman-akib) [inline src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle.java:840]: Use `Util.fixEmptyAndTrim()`

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle.java [hunk 2]
```diff
@@ -733,6 +835,76 @@ public void setMark_build_status(String mark_build_status) {
...
+    }
+
+    @DataBoundSetter
+    public void setSrm_apikey(String srm_apikey) {
+        this.srm_apikey = srm_apikey;
```
Comment (zaman-akib) [inline src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle.java:845]: Use `Util.fixEmptyAndTrim()`

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle.java [hunk 3]
```diff
@@ -733,6 +835,76 @@ public void setMark_build_status(String mark_build_status) {
...
+    }
+
+    @DataBoundSetter
+    public void setSrm_branch_parent(String srm_branch_parent) {
+        this.srm_branch_parent = Util.fixEmptyAndTrim(srm_branch_parent);
+    }
+
+    @DataBoundSetter
+    public void setSrm_sca_search_depth(Integer srm_sca_search_depth) {
+        this.srm_sca_search_depth = srm_sca_search_depth;
```
Comment (zaman-akib) [inline src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle.java:875]: Use `Util.fixEmptyAndTrim()` here and check other setters

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/pipeline/SecurityScanStep.java [hunk 4]
```diff
@@ -790,6 +860,41 @@ public void setPolaris_reports_sarif_issue_types(String polaris_reports_sarif_is
         this.polaris_reports_sarif_issue_types = Util.fixEmptyAndTrim(polaris_reports_sarif_issue_types);
     }
 
+    @DataBoundSetter
+    public void setSrm_url(String srm_url) {
+        this.srm_url = srm_url;
```
Comment (zaman-akib) [inline src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/pipeline/SecurityScanStep.java:865]: Use `Util.fixEmptyAndTrim()`

File: src/main/resources/io/jenkins/plugins/synopsys/security/scan/extension/pipeline/SecurityScanStep/help-srm_project_id.html [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<div>
+    Project id in SRM server.
```
Comment (zaman-akib) [inline src/main/resources/io/jenkins/plugins/synopsys/security/scan/extension/pipeline/SecurityScanStep/help-srm_project_id.html:2]: Project ID in SRM server

File: src/main/resources/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle/help-srm_project_id.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<div>
+    Project id in SRM server.
```
Comment (zaman-akib) [inline src/main/resources/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle/help-srm_project_id.html:2]: Project ID in SRM server.

**Review by zaman-akib — CHANGES_REQUESTED** [review]

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle.java [hunk 7]
```diff
@@ -85,6 +87,22 @@ public class SecurityScanFreestyle extends Builder implements SecurityScan, Free
     private String polaris_sast_config_path;
     private String polaris_sast_args;
 
+    private String srm_url;
+    private String srm_apikey;
```
Comment (zaman-akib) [inline src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle.java:91]: Use `transient` keyword 

`private transient String srm_apikey;`

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/pipeline/SecurityScanStep.java [hunk 8]
```diff
@@ -112,6 +114,15 @@ public class SecurityScanStep extends Step implements SecurityScan, PrCommentSca
     private String coverity_project_directory;
     private String blackduck_project_directory;
     private String polaris_project_directory;
+    private String srm_project_directory;
+
+    private String srm_url;
+    private String srm_apikey;
```
Comment (zaman-akib) [inline src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/pipeline/SecurityScanStep.java:120]: Use `transient` keyword 

`private transient String srm_apikey;`

**Review by zaman-akib — CHANGES_REQUESTED** [review]

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/input/srm/SRM.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class SRM {
+    @SuppressWarnings("lgtm[jenkins/plaintext-storage]")
+    @JsonProperty("url")
+    private String url;
+
+    @JsonProperty("apikey")
+    private String apikey;
```
Comment (zaman-akib) [inline src/main/java/io/jenkins/plugins/synopsys/security/scan/input/srm/SRM.java:11]: Use `@SuppressWarnings("lgtm[jenkins/plaintext-storage]")` above `@JsonProperty("apikey")`

**Review by zaman-akib — CHANGES_REQUESTED** [review]

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/global/ScannerGlobalConfig.java [hunk 10]
```diff
@@ -255,6 +296,10 @@ public ListBoxModel doFillCoverityCredentialsIdItems() {
                         ScanCredentialsHelper.USERNAME_PASSWORD_CREDENTIALS);
     }
 
+    public ListBoxModel doFillSrmCredentialsIdItems() {
```
Comment (zaman-akib) [inline src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/global/ScannerGlobalConfig.java:299]: Use `@SuppressWarnings({"lgtm[jenkins/no-permission-check]", "lgtm[jenkins/csrf]"})`

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle.java (hunk 1 above)
Comment (sadmananik) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/extension/freestyle/SecurityScanFreestyle.java:840]: I saw many fields missing fixEmptyAndTrim that are unavailable in the pipeline syntax. So it's fine.

**Review by zaman-akib — APPROVED** [review]

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java [hunk 11]
```diff
@@ -253,6 +273,56 @@ private void setScanObject(
...
+        if (scanParameters.containsKey(ApplicationConstants.SRM_SAST_EXECUTION_PATH_KEY)) {
+            String installationPath = scanParameters
+                    .get(ApplicationConstants.SRM_SAST_EXECUTION_PATH_KEY)
+                    .toString()
+                    .trim();
+            if (installationPath != null && !installationPath.isBlank()) {
+                if (coverity == null) coverity = new Coverity();
+                io.jenkins.plugins.synopsys.security.scan.input.coverity.Execution execution =
+                        new io.jenkins.plugins.synopsys.security.scan.input.coverity.Execution();
+                execution.setPath(installationPath);
```
Comment (jahid1209) [inline src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java:322]: Same here!

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java [hunk 12]
```diff
@@ -253,6 +273,56 @@ private void setScanObject(
...
+        if (scanParameters.containsKey(ApplicationConstants.SRM_SCA_EXECUTION_PATH_KEY)) {
+            String installationPath = scanParameters
+                    .get(ApplicationConstants.SRM_SCA_EXECUTION_PATH_KEY)
+                    .toString()
+                    .trim();
+            if (installationPath != null && !installationPath.isBlank()) {
+                if (blackDuck == null) blackDuck = new BlackDuck();
+                io.jenkins.plugins.synopsys.security.scan.input.blackduck.Execution execution =
+                        new io.jenkins.plugins.synopsys.security.scan.input.blackduck.Execution();
+                execution.setPath(installationPath);
```
[the remark this answers is not in the document]
Comment (sadmananik) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java:303]: There are two Class with same "Execution" name one for blackduck and one for coverity

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java (hunk 11 above)
Comment (sadmananik) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java:322]: There are two Class with same "Execution" name one for blackduck and one for coverity

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/PluginParametersHandler.java [hunk 13]
```diff
@@ -126,33 +126,50 @@ private void logMessagesForProductParameters(Map<String, Object> scanParameters,
...
+    private void logFilteredParameters(Map<String, Object> filteredParameters, String securityProduct) {
+        for (Map.Entry<String, Object> entry : filteredParameters.entrySet()) {
+            String key = entry.getKey();
+            Object value = entry.getValue();
+
+            if (shouldLogParameter(securityProduct, key)) {
+                if (isSensitiveKey(key)) {
+                    value = LogMessages.ASTERISKS;
+                }
+                logger.info(LogMessages.LOG_DASH + key + " = " + value.toString());
```
[the remark this answers is not in the document]
Comment (sadmananik) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/PluginParametersHandler.java:146]: It was like this before. Also, While using String.valueOf(value) it shows unnecessary code in sonarLint.

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/service/scan/polaris/PolarisParametersService.java [hunk 14]
```diff
@@ -57,43 +63,51 @@ private List<String> getInvalidPolarisParamsForAllJobTypes(Map<String, Object> p
...
+
+            logger.error(
+                    missingMandatoryParamsForAllJobTypes + " is mandatory parameter for " + jobTypeName + " job type");
         }
 
-        return invalidParamsForPipelineOrFreeStyle;
+        return missingMandatoryParamsForAllJobTypes;
+    }
+
+    private List<String> getFreeStyleAndPipelinePolarisMissingMandatoryParams(Map<String, Object> polarisParameters) {
```
[the remark this answers is not in the document]
Comment (sadmananik) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/service/scan/polaris/PolarisParametersService.java:93]: ok

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/service/scan/srm/SRMParametersService.java [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+                jobTypeName = "Pipeline";
+            }
+
+            logger.error(missingMandatoryParams + " is mandatory parameter for " + jobTypeName + " job type");
+        }
+
+        return missingMandatoryParams;
+    }
+
+    private List<String> getFreeStyleAndPipelineSrmMissingMandatoryParams(Map<String, Object> srmParameters) {
```
[the remark this answers is not in the document]
Comment (sadmananik) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/service/scan/srm/SRMParametersService.java:87]: ok

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jahid1209) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java:303]: Then import the appropriate one

[the remark this answers is not in the document]
Comment (sadmananik) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java:303]: I need both here.

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/PluginParametersHandler.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (sadmananik) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/PluginParametersHandler.java:146]: No. IDE also suggested it as unnecessary.

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jahid1209) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java:303]: Then import one and use the fully qualified name for the other.

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/PluginParametersHandler.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (jahid1209) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/PluginParametersHandler.java:146]: We should keep it as it is null safe

**Review by maksudur-rahman-maruf — APPROVED** [review]

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (sadmananik) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/service/ScannerArgumentService.java:303]: Whats the problem with the current implementation?

File: src/main/java/io/jenkins/plugins/synopsys/security/scan/PluginParametersHandler.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (sadmananik) [reply src/main/java/io/jenkins/plugins/synopsys/security/scan/PluginParametersHandler.java:146]: Yes, I don't know why we missed it the first time.

**Pull request merged**